src/Nethermind/Nethermind.Blockchain/FullPruning/FullPruner.cs

```diff
@@ -217,49 +236,40 @@ protected virtual void RunPruning(IPruningContext pruning, Hash256 statRoot)
                     writeFlags |= WriteFlags.LowPriority;
                 }
 
-                using CopyTreeVisitor copyTreeVisitor = new(pruning, writeFlags, _logManager);
+                using CopyTreeVisitor copyTreeVisitor = new(pruning, cancellationToken, writeFlags, _logManager);
                 VisitingOptions visitingOptions = new()
                 {
                     MaxDegreeOfParallelism = _pruningConfig.FullPruningMaxDegreeOfParallelism,
                     FullScanMemoryBudget = ((long)_pruningConfig.FullPruningMemoryBudgetMb).MiB(),
                 };
                 if (_logger.IsInfo) _logger.Info($"Full pruning started with MaxDegreeOfParallelism: {visitingOptions.MaxDegreeOfParallelism} and FullScanMemoryBudget: {visitingOptions.FullScanMemoryBudget}");
-                _stateReader.RunTreeVisitor(copyTreeVisitor, statRoot, visitingOptions);
+                _stateReader.RunTreeVisitor(copyTreeVisitor, stateRoot, visitingOptions);
 
-                if (!pruning.CancellationTokenSource.IsCancellationRequested)
+                if (!cancellationToken.IsCancellationRequested)
                 {
-                    void CommitOnNewBLock(object o, OnUpdateMainChainArgs e)
+                    copyTreeVisitor.Finish();
+
+                    await WaitForMainChainChange(cancellationToken, (e) =>
                     {
-                        if (!e.WereProcessed) return;
-                        _blockTree.OnUpdateMainChain -= CommitOnNewBLock;
-                        // ReSharper disable AccessToDisposedClosure
+                        // The db swap happens here. We do it within the event handler of main chain change to block
+                        // so that it does not happen during block processing.
                         pruning.Commit();
-                        _lastPruning = DateTime.Now;
-                        pruning.Dispose();
-                        // ReSharper restore AccessToDisposedClosure
-                    }
+                        return true;
+                    });
 
-                    _blockTree.OnUpdateMainChain += CommitOnNewBLock;
-                    copyTreeVisitor.Finish();
-                }
-                else
-                {
-                    pruning.Dispose();
+                    _lastPruning = DateTime.Now;
```

Comment:
Really should use `DateTime.UtcNow` for system type dates; .Now can do weird things on time changes (daylight savings, timezone changes if laptop)

Reply:
Done